client_code/persistence.py

```diff
@@ -49,86 +49,103 @@ def __set__(self, instance, value):
         instance._delta[self._name] = value
 
 
-@classmethod
-def _search(cls, *args, **kwargs):
-    rows = anvil.server.call(f"search_{cls.__name__.lower()}", *args, **kwargs)
-    return (cls.create(store=row) for row in rows)
-
-
-@classmethod
-def _create(cls, store=None, delta=None, *args, **kwargs):
-    instance = cls(*args, **kwargs)
-    instance._store = store or {}
-    instance._delta = delta or {}
-    return instance
+class LinkedClass:
+    "A descriptor class for adding objects based on linked tables as attributes"
 
+    def __init__(self, cls, linked_column=None, *args, **kwargs):
+        self._linked_column = linked_column
+        self._cls = cls
+        self._obj = None
+        self._args = args or []
+        self._kwargs = kwargs or {}
 
-def _getattr(self, key):
-    if self._delta and key in self._delta:
-        return self._delta[key]
+    def __set_name__(self, owner, name):
+        if self._linked_column is None:
+            self._linked_column = name
 
-    return dict(self._store).get(key, None)
+    def __get__(self, instance, objtype=None):
+        if instance is None:
+            return self
 
+        if self._obj is None:
+            self._obj = self._cls(
+                instance._store[self._linked_column], *self._args, **self._kwargs
+            )
 
-def _getitem(self, key):
-    return getattr(self, key)
+        return self._obj
 
+    def __set__(self, instance, value):
+        raise ValueError("Linked Class instance is already set and cannot be changed")
 
-def _set_value(self, key, value):
-    is_private = key.startswith("_")
-    is_descriptor = hasattr(self.__class__, key) and hasattr(
-        getattr(self.__class__, key), "__set__"
-    )
-    if is_private or is_descriptor:
-        object.__setattr__(self, key, value)
-    else:
-        self._delta[key] = value
 
+class PersistedClass:
+    @classmethod
+    def search(cls, *args, **kwargs):
+        rows = anvil.server.call(f"search_{cls.__name__.lower()}", *args, **kwargs)
+        return (cls.create(store=row) for row in rows)
 
-def _class_name(instance):
-    return "".join(
-        "_" + c.lower() if c.isupper() else c for c in instance.__class__.__name__
-    ).lstrip("_")
+    def __init__(self, store=None, delta=None, *args, **kwargs):
+        self._store = store or {}
+        self._delta = delta or {}
```

Comment:
do you see a use case for a custom `store`/`delta`?
I'm wondering if it's better to remove this api and include it in the future if we need it.

Alternative api suggestion: use a `store_factory` and `delta_factory` function

```python
class PersistedClass:
    _store_factory = dict
    _delta_factory = dict

    def __init__(self, *args, **kwargs):
        self._store = self._store_factory()
        self._delta = self._delta_factory()

```

Reply:
Yes - that's how I'm thinking of handling local storage in IndexedDb with syncing to server.

Reply:
Can we keep it as is for now and deal with improving it when we tackle the offline sync stuff?

Reply:
I can see some argument for getting rid of the delta parameter, but the store needs to stay so you can pass a data tables row to the init.

Reply:
e.g. In retrofitting this to an existing app, I have situations where I already have the row and need to create an instance.